File: Algorithms/0877_Stone_Game/Java/StoneGameSolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    private Map<String, Boolean> cache;

    public boolean stoneGame(int[] piles) {
        if (piles.length <= 2) return true;
        cache = new HashMap<>();
        return dfs(0, piles.length - 1, piles, 0, 0, 1);
    }

    private boolean dfs(int left, int right, int[] piles, int alex, int lee, int flag) {
        if (left > right) return alex > lee;
        String key = left + "," + right;
        if (cache.containsKey(key)) return cache.get(key);
        boolean res = false;

        // flag==1: alex turn
        // flag==0: lee turn
        if (flag == 1) {
            alex += piles[left];
            res = dfs(left + 1, right, piles, alex, lee, 0);
            alex -= piles[left];
            cache.put(key, res);
            if (res) return res;

            alex += piles[right];
            res = dfs(left, right - 1, piles, alex, lee, 0);
            alex -= piles[right];
            cache.put(key, res);
            if (res) return res;
        }
        if (flag == 0) {
            lee += piles[left];
            res = dfs(left + 1, right, piles, alex, lee, 1);
            lee -= piles[left];
            cache.put(key, res);
            if (res) return res;

            lee += piles[right];
            res = dfs(left, right - 1, piles, alex, lee, 1);
            lee -= piles[right];
            cache.put(key, res);
            if (res) return res;
        }
        return res;
    }
}

File: Algorithms/0877_Stone_Game/Python/Stone_Game_Solution_1.py

# Space: O(n)
# Time: O(n)

class Solution:
    def stoneGame(self, piles):
        length = len(piles)
        if length <= 2: return True

        cache = {}

        # flag==1: alex turn
        # flag==0: lee turn
        def dfs(left, right, nums, alex, lee, flag):
            if left > right: return alex > lee
            key = str(left) + ',' + str(right)
            if key in cache: return cache[key]

            if flag == 1:
                alex += nums[left]
                res = dfs(left + 1, right, nums, alex, lee, 0)
                cache[key] = res
                alex -= nums[left]
                if res: return res

                alex += nums[right]
                res = dfs(left, right - 1, nums, alex, lee, 0)
                cache[key] = res
                alex -= nums[right]
                if res: return res

            if flag == 0:
                lee += nums[left]
                res = dfs(left + 1, right, nums, alex, lee, 1)
                cache[key] = res
                lee -= nums[left]
                if res: return res

                lee += nums[right]
                res = dfs(left, right - 1, nums, alex, lee, 1)
                cache[key] = res
                lee -= nums[right]
                if res: return res

            return False

        return dfs(0, length - 1, piles, 0, 0, 1)





